packages/gatsby/src/utils/page-data.ts

```diff
@@ -0,0 +1,79 @@
+import fs from "fs-extra"
+import path from "path"
+import { IGatsbyPage } from "../redux/types"
+import { store } from "../redux"
+
+interface IPageDataContext {
+  publicDir: string
+}
+
+interface IPageData
+  extends Pick<IGatsbyPage, "componentChunkName" | "matchPath" | "path"> {
+  result: any
+}
+
+export const fixedPagePath = (pagePath: string): string =>
+  pagePath === `/` ? `index` : pagePath
+
+const getFilePath = (
+  { publicDir }: IPageDataContext,
+  pagePath: string
+): string =>
+  path.join(publicDir, `page-data`, fixedPagePath(pagePath), `page-data.json`)
+
+export const read = async (
+  { publicDir }: IPageDataContext,
+  pagePath: string
+): Promise<IPageData> => {
+  const filePath = getFilePath({ publicDir }, pagePath)
+  const rawPageData = await fs.readFile(filePath, `utf-8`)
+
+  return JSON.parse(rawPageData)
+}
+
+export const remove = async (
+  { publicDir }: IPageDataContext,
+  pagePath: string
+): Promise<void> => {
+  const filePath = getFilePath({ publicDir }, pagePath)
+
+  if (fs.existsSync(filePath)) {
+    return await fs.remove(filePath)
+  }
+
+  return Promise.resolve()
+}
+
+export const write = async (
+  { publicDir }: IPageDataContext,
+  { componentChunkName, matchPath, path }: Exclude<IPageData, "result">,
```

Comment:
I don't like this Exclude. We try to avoid complex typing. I think it would be easier to just have two interfaces. One that extends the other and adds result.